src/graph.rs

```diff
@@ -232,37 +266,86 @@ impl ModuleSlot {
   }
 }
 
+/// Convert a module slot entry into a result which contains the resolved
+/// module specifier and media type or the module graph error.
+#[cfg(feature = "rust")]
+fn to_result(
+  (specifier, module_slot): (&ModuleSpecifier, &ModuleSlot),
+) -> Option<(
+  ModuleSpecifier,
+  Result<(ModuleSpecifier, MediaType), ModuleGraphError>,
+)> {
+  match module_slot {
+    ModuleSlot::Err(err) => Some((specifier.clone(), Err(err.clone()))),
+    ModuleSlot::Module(module) => Some((
+      specifier.clone(),
+      Ok((module.specifier.clone(), module.media_type)),
+    )),
+    _ => None,
+  }
+}
+
+/// The structure which represents a module graph, which can be serialized as
+/// well as "printed".  The roots of the graph represent the "starting" point
+/// which can be located in the module "slots" in the graph. The graph also
+/// contains any redirects where the requested module specifier was redirected
+/// to another module specifier when being loaded.
 #[derive(Debug, Serialize)]
 pub struct ModuleGraph {
-  pub root: ModuleSpecifier,
+  pub roots: Vec<ModuleSpecifier>,
   #[serde(skip_serializing)]
-  maybe_locker: Option<Rc<RefCell<dyn Locker>>>,
-  #[serde(serialize_with = "serialize_modules")]
-  pub(crate) modules: BTreeMap<ModuleSpecifier, ModuleSlot>,
-  redirects: BTreeMap<ModuleSpecifier, ModuleSpecifier>,
+  maybe_locker: Option<Rc<RefCell<Box<dyn Locker>>>>,
```

Comment:
`Rc<RefCell<Box<dyn Locker>>` looks like antipattern and I'm surprised clippy doesn't complain. I think you should be able to store it in `Rc<RefCell<dyn Locker>>` instead.

Reply:
Yes, I was hacking away on this so much that I didn't realise what I ended up with.

Reply:
Hmmm... no, this doesn't work at the other end... and I can't cast it to a locker unless the trait is sized, but trying to add sized to a trait is beyond me and causes a whole lot of other issues.